restcomm/restcomm.http/src/main/java/org/restcomm/connect/http/asyncclient/HttpAsycClientHelper.java

```diff
@@ -0,0 +1,236 @@
+        HttpUriRequest request = null;
+        HttpRequestDescriptor temp = descriptor;
+        try {
+
+            request = request(temp);
+            request.setHeader("http.protocol.content-charset", "UTF-8");
+            if (descriptor.getTimeout() > 0){
+                HttpContext httpContext = new BasicHttpContext();
+                httpContext.setAttribute(HttpClientContext.REQUEST_CONFIG, RequestConfig.custom().
+                setConnectTimeout(descriptor.getTimeout()).
+                setSocketTimeout(descriptor.getTimeout()).
+                setConnectionRequestTimeout(descriptor.getTimeout()).build());
+                client.execute(request, httpContext, getFutureCallback(request, sender));
+            } else {
+                client.execute(request, getFutureCallback(request, sender));
+            }
+        } catch (Exception e) {
+            logger.error("Problem while trying to execute http request {}, exception: {}", request.getRequestLine(), e);
+            throw e;
+        }
+    }
+
+    private FutureCallback<HttpResponse> getFutureCallback(final HttpUriRequest request, final ActorRef sender){
+        return new FutureCallback<HttpResponse>(){
+
+            @Override
+            public void completed(HttpResponse result) {
+                if (logger.isDebugEnabled()) {
+                    logger.debug(String.format("success on execution of http request. result %s", result));
+                }
+                DownloaderResponse response = null;
+                try {
+                    response = new DownloaderResponse(response(request, result));
+                } catch (IOException e) {
+                    logger.error("Exception while parsing response", e);
+                    response = new DownloaderResponse(e, "Exception while parsing response");
+                }
+                sendResponse(sender, response);
+            }
+
+            @Override
+            public void failed(Exception ex) {
+                logger.error("got failure on executing http request {}, exception: {}", request.getRequestLine(), ex);
+                DownloaderResponse response = new DownloaderResponse(ex, "got failure on executing http request");
+                sendResponse(sender, response);
+            }
+
+            @Override
+            public void cancelled() {
+                logger.warning("got cancellation on executing http request {}", request.getRequestLine());
+                DownloaderResponse response = new DownloaderResponse(new Exception("got cancellation on executing http request"), "got cancellation on executing http request");
+                sendResponse(sender, response);
+            }};
+    }
+
+    private void sendResponse(final ActorRef sender, Object response){
+        if (sender != null && !sender.isTerminated()) {
+            sender.tell(response, self());
+        } else {
+            if (logger.isInfoEnabled()) {
```

Comment:
what happens if we send msg to terminated actor? checking actor state from the outside is not guaranteed,as in the next instruction the state may have changed, so there will be chances where we send the response,and the actor is terminated, not sure is worth to make the distinction here, or maybe send response blindly to actor, and let Akka figure out what to do (deadletter box..)

Reply:
agree with letting akka handle the delivery dynamics